# **PROJECT 1 - PART 5**
*   Samar KRIMI
*   03/08/2023
---

# **Part 1**

In [197]:
# Import the data as a Pandas DataFrame
import pandas as pd

In [198]:
# Load data 'sales_predictions_2023.csv'
filename = '/content/sales_predictions_2023.csv'

In [199]:
# Read data
df = pd.read_csv(filename)
df

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDA15        9.300          Low Fat         0.016047   
1              DRC01        5.920          Regular         0.019278   
2              FDN15       17.500          Low Fat         0.016760   
3              FDX07       19.200          Regular         0.000000   
4              NCD19        8.930          Low Fat         0.000000   
...              ...          ...              ...              ...   
8518           FDF22        6.865          Low Fat         0.056783   
8519           FDS36        8.380          Regular         0.046982   
8520           NCJ29       10.600          Low Fat         0.035186   
8521           FDN46        7.210          Regular         0.145221   
8522           DRG01       14.800          Low Fat         0.044878   

                  Item_Type  Item_MRP Outlet_Identifier  \
0                     Dairy  249.8092            OUT049   
1               Soft Drinks   48.2692            OUT018   
2                      Meat  141.6180            OUT049   
3     Fruits and Vegetables  182.0950            OUT010   
4                 Household   53.8614            OUT013   
...                     ...       ...               ...   
8518            Snack Foods  214.5218            OUT013   
8519           Baking Goods  108.1570            OUT045   
8520     Health and Hygiene   85.1224            OUT035   
8521            Snack Foods  103.1332            OUT018   
8522            Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1999      Medium               Tier 1   
1                          2009      Medium               Tier 3   
2                          1999      Medium               Tier 1   
3                          1998         NaN               Tier 3   
4                          1987        High               Tier 3   
...                         ...         ...                  ...   
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
0     Supermarket Type1          3735.1380  
1     Supermarket Type2           443.4228  
2     Supermarket Type1          2097.2700  
3         Grocery Store           732.3800  
4     Supermarket Type1           994.7052  
...                 ...                ...  
8518  Supermarket Type1          2778.3834  
8519  Supermarket Type1           549.2850  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700  

[8523 rows x 12 columns]

In [200]:
# The top 5 rows of the dataframe
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

# **Part 2**

In [201]:
# Examine the number of unique values in columns
df.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

**Question 1 :** How many rows and columns

In [202]:
# Obtain essential information about the data : check columns, rows, data types, and missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [203]:
df.shape

(8523, 12)

*   8523 rows and 12 columns



**Question 2 :** The datatypes of each variable

In [204]:
# Get the data type for all columns
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

*   7 objects, 4 floats, 1 integer



**Question 3  :** Find any duplicates and drop them

In [205]:
# Find duplicated
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
8518    False
8519    False
8520    False
8521    False
8522    False
Length: 8523, dtype: bool

In [206]:
# check duplicated data
df.duplicated().any()

False

In [207]:
# Count the duplicates
df.duplicated().sum()

0

*   there is no duplicated data

**Question 4 :** Identify missing values

In [208]:
# Check the df to see if there are missing values : Sums the number of nas in each column
#df.isna().sum()
null_vals=df.isna().sum()
null_vals

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

*  For variable Item_Weight there are 1463 nas (8523-7060)
*  For variable Outlet_Size there are 2410 nas (8523-6113)

**Question 5 & 6 :** Decide on how to address the missing values & Confirm that there are no missing values after addressing them

In [209]:
# Check the Item_Weight column
df['Item_Weight'].value_counts()

12.150    86
17.600    82
13.650    77
11.800    76
15.100    68
          ..
7.275      2
7.685      1
9.420      1
6.520      1
5.400      1
Name: Item_Weight, Length: 415, dtype: int64

In [210]:
# check the mean value of Item_Weight variable
df[['Item_Weight']].mean()

Item_Weight    12.857645
dtype: float64

In [211]:
# Imputing Missing Data of Item_Weight column by its mean value
df['Item_Weight'].fillna(value = 12.857, inplace = True)

In [212]:
# There are no missing values after addressing Item_Weight column
df.isna().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [213]:
# Check the categorical values of Outlet_Size variable
df['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [214]:
# Imputing Missing Data of Outlet_Size column by its categorical value equal to Medium
df['Outlet_Size'].fillna(value='Medium', inplace = True)

In [215]:
# There are no missing values after addressing Outlet_Size column
df.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

* Imputation with **mean value** for variable Item_Weight : it contains only values
* Imputation with **Medium** for variable Outlet_Size : the most repetitive  category

**Question 7 :** Find and fix any inconsistent categories of data this can include misspellings, abbreviations or typos

In [216]:
# Get all datatypes
data_types = df.dtypes
data_types

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

In [217]:
# Get all object datatypes
str_cols = data_types[data_types=="object"]
str_cols

Item_Identifier         object
Item_Fat_Content        object
Item_Type               object
Outlet_Identifier       object
Outlet_Size             object
Outlet_Location_Type    object
Outlet_Type             object
dtype: object

* There are 7 objects : Item_Identifier, Item_Fat_Content, Item_Type, Outlet_Identifier, Outlet_Size, Outlet_Location_Type, Outlet_Type

In [218]:
# We use .value_counts() to examine Item_Identifier column
df['Item_Identifier'].value_counts()

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64

In [219]:
# We use .value_counts() to examine Item_Fat_Content column
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [220]:
# Inconsistent categories of data from Item_Fat_Content column are fixed
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace(['LF', 'low fat','reg'],['Low Fat','Low Fat','Regular'])
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

* Inconsistent categories of data from Item_Fat_Content are fixed : LF and low fat are replaced by Low Fat and reg is replaced by Regular


In [221]:
# We use .value_counts() to examine Item_Type column
df['Item_Type'].value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

* Household and Health and Hygiene are unrelated to Item_type column that
describes the category to which the product belongs there must be replaced by others

In [222]:
#  Inconsistent categories of data from Item_Type column are fixed
df['Item_Type'] = df['Item_Type'].replace(['Household', 'Health and Hygiene'],['Others','Others'])
df['Item_Type'].value_counts()

Others                   1599
Fruits and Vegetables    1232
Snack Foods              1200
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

In [223]:
# We use .value_counts() to examine Outlet_Identifier column
df['Outlet_Identifier'].value_counts()

OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64

In [224]:
# We use .value_counts() to examine Outlet_Size column
df['Outlet_Size'].value_counts()

Medium    5203
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [225]:
# We use .value_counts() to examine Outlet_Location_Type column
df['Outlet_Location_Type'].value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

In [226]:
# We use .value_counts() to examine Outlet_Size column
df['Outlet_Type'].value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

**Question 8 :** Obtain the summary statistics of each numerical columns

In [227]:
# Get all float64 datatypes
data_types[data_types=="float64"]

Item_Weight          float64
Item_Visibility      float64
Item_MRP             float64
Item_Outlet_Sales    float64
dtype: object

In [228]:
# Get all int64 datatypes
data_types[data_types=="int64"]

Outlet_Establishment_Year    int64
dtype: object

In [229]:
# Show summary stats for all numeric columns
df.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  8523.000000      8523.000000  8523.000000                8523.000000   
mean     12.857534         0.066132   140.992782                1997.831867   
std       4.226124         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       9.310000         0.026989    93.826500                1987.000000   
50%      12.857000         0.053931   143.012800                1999.000000   
75%      16.000000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

In [230]:
# find summary stats for each numerical column and round to two decimal places
df.describe().round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
count      8523.00          8523.00   8523.00                    8523.00   
mean         12.86             0.07    140.99                    1997.83   
std           4.23             0.05     62.28                       8.37   
min           4.56             0.00     31.29                    1985.00   
25%           9.31             0.03     93.83                    1987.00   
50%          12.86             0.05    143.01                    1999.00   
75%          16.00             0.09    185.64                    2004.00   
max          21.35             0.33    266.89                    2009.00   

       Item_Outlet_Sales  
count            8523.00  
mean             2181.29  
std              1706.50  
min                33.29  
25%               834.25  
50%              1794.33  
75%              3101.30  
max             13086.96

In [231]:
# To View if there's a space in the columns name, no inconsistence in this categorial feature
df.columns

Index(['Item_Identifier', 'Item_Weight', 'Item_Fat_Content', 'Item_Visibility',
       'Item_Type', 'Item_MRP', 'Outlet_Identifier',
       'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type',
       'Outlet_Type', 'Item_Outlet_Sales'],
      dtype='object')

# **Part 5**

In [232]:
# Import all librairies
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer, make_column_selector
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.impute import SimpleImputer
from sklearn.pipeline import make_pipeline, Pipeline
# Set the default transformation output to pandas
from sklearn import set_config
set_config(transform_output='pandas')

In [233]:
# Get info for cleaned dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                8523 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                8523 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Check the categorical values of object features

In [234]:
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace(['Low Fat'],['Low'])

In [235]:
df['Item_Fat_Content'].value_counts()

Low        5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [236]:
df['Outlet_Size'].value_counts()

Medium    5203
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [237]:
df['Outlet_Location_Type'].value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

In [238]:
df['Outlet_Type'] = df['Outlet_Type'].replace(['Grocery Store'],['Supermarket Type4'])

In [239]:
df['Outlet_Type'].value_counts()

Supermarket Type1    5577
Supermarket Type4    1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

In [240]:
df['Outlet_Identifier'].value_counts()

OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64

In [241]:
df['Item_Type'].value_counts()

Others                   1599
Fruits and Vegetables    1232
Snack Foods              1200
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

**Identify the features (X) and target (y)**

In [242]:
# Drop 'Item_Identifier', 'Outlet_Identifier' & 'Item_Type' also because they're not relevant for this study
# split X and y, you are predicting rating
X = df.drop(columns=['Item_Outlet_Sales','Item_Identifier', 'Outlet_Identifier', 'Item_Type'])
y = df['Item_Outlet_Sales']
# Split training and test, set random_state to 42 for reproducibility
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [243]:
# Showing X_train and y_train
display(X_train.head(), y_train.head())

Item_Weight Item_Fat_Content  Item_Visibility  Item_MRP  \
4776       16.350              Low         0.029565  256.4646   
7510       15.250          Regular         0.000000  179.7660   
5828       12.350          Regular         0.158716  157.2946   
5327        7.975              Low         0.014628   82.3250   
4810       19.350              Low         0.016645  120.9098   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
4776                       2009      Medium               Tier 3   
7510                       2009      Medium               Tier 3   
5828                       1999      Medium               Tier 1   
5327                       2004       Small               Tier 2   
4810                       2002      Medium               Tier 2   

            Outlet_Type  
4776  Supermarket Type2  
7510  Supermarket Type2  
5828  Supermarket Type1  
5327  Supermarket Type1  
4810  Supermarket Type1

4776     515.3292
7510    3056.0220
5828    1577.9460
5327    1331.6000
4810    1687.1372
Name: Item_Outlet_Sales, dtype: float64

In [244]:
# Showing X_test and y_test
display(X_test.head(), y_test.head())

Item_Weight Item_Fat_Content  Item_Visibility  Item_MRP  \
7503       14.300              Low         0.026300   79.4302   
2957        7.930              Low         0.071136   42.7086   
7031       14.500          Regular         0.041313   42.0454   
1084       12.857          Regular         0.044767  173.7054   
856        10.195          Regular         0.012456  197.5110   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
7503                       1987        High               Tier 3   
2957                       1997       Small               Tier 1   
7031                       1999      Medium               Tier 1   
1084                       1985      Medium               Tier 3   
856                        2004       Small               Tier 2   

            Outlet_Type  
7503  Supermarket Type1  
2957  Supermarket Type1  
7031  Supermarket Type1  
1084  Supermarket Type3  
856   Supermarket Type1

7503    1743.0644
2957     356.8688
7031     377.5086
1084    5778.4782
856     2356.9320
Name: Item_Outlet_Sales, dtype: float64

# **Preprocessing Pipeline to prepare the dataset for ML**

In [245]:
# Check types after split : 4 categorical features
X.dtypes

Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_MRP                     float64
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
dtype: object

In [262]:
# Creating a categorical data selector
num_selector = make_column_selector(dtype_include = 'number')
num_selector(X_train)

['Item_Weight', 'Item_Visibility', 'Item_MRP', 'Outlet_Establishment_Year']

In [246]:
# Creating a categorical data selector
cat_selector = make_column_selector(dtype_include = 'object')
cat_selector(X_train)

['Item_Fat_Content', 'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type']

Checking unique values in categorical features

In [247]:
# Checking unique values in 'Item_Fat_Content' feature : nominal feature
df['Item_Fat_Content'].unique()

array(['Low', 'Regular'], dtype=object)

In [248]:
# Checking unique values in 'Outlet_Size' feature : ordinal feature
df['Outlet_Size'].unique()

array(['Medium', 'High', 'Small'], dtype=object)

In [249]:
# Checking unique values in 'Outlet_Location_Type' feature : nominal feature
df['Outlet_Location_Type'].unique()

array(['Tier 1', 'Tier 3', 'Tier 2'], dtype=object)

In [250]:
# Checking unique values in 'Outlet_Type' feature : nominal feature
df['Outlet_Type'].unique()

array(['Supermarket Type1', 'Supermarket Type2', 'Supermarket Type4',
       'Supermarket Type3'], dtype=object)

Categorical features : 3 Nominal, 1 ordinal

**Create Pipelines and Define Tuples**

Ordinal feature pipeline

In [251]:
# PREPROCESSING PIPELINE FOR ORDINAL DATA
# Save list of column names
ord_cols = ['Outlet_Size']
print("Ordinal Columns:", ord_cols)
# Create imputer for ordinal data
impute_ord = SimpleImputer(strategy='most_frequent')
## Making the OrdinalEncoder
# Specifying order of categories for our  Ordinal Qual/Cond Columms
size_order = ['High','Medium', 'Small']
# Making the list of order lists for OrdinalEncoder
ordinal_category_orders = [size_order]
ord_encoder = OrdinalEncoder(categories=ordinal_category_orders)
# Making a final scaler to scale category #'s
scaler_ord = StandardScaler()
## Making an ord_pipe
ord_pipe = make_pipeline(impute_ord, ord_encoder, scaler_ord)
ord_pipe

Ordinal Columns: ['Outlet_Size']


Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='most_frequent')),
                ('ordinalencoder',
                 OrdinalEncoder(categories=[['High', 'Medium', 'Small']])),
                ('standardscaler', StandardScaler())])

In [252]:
# Making an ordinal_tuple for ColumnTransformer
ord_tuple = ('ordinal', ord_pipe, ord_cols)
ord_tuple

('ordinal',
 Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='most_frequent')),
                 ('ordinalencoder',
                  OrdinalEncoder(categories=[['High', 'Medium', 'Small']])),
                 ('standardscaler', StandardScaler())]),
 ['Outlet_Size'])

Nominal feature pipeline

In [253]:
# PREPROCESSING PIPELINE FOR ONE-HOT-ENCODED DATA
# Save list of column names
ohe_cols = X_train.select_dtypes('object').drop(columns=ord_cols).columns
print("OneHotEncoder Columns:", ohe_cols)
# Instantiate the individual preprocessors
impute_ohe = SimpleImputer(strategy='constant')
ohe_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
# Make pipeline with imputer and encoder
ohe_pipe = make_pipeline(impute_ohe, ohe_encoder)
ohe_pipe

OneHotEncoder Columns: Index(['Item_Fat_Content', 'Outlet_Location_Type', 'Outlet_Type'], dtype='object')


Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='constant')),
                ('onehotencoder',
                 OneHotEncoder(handle_unknown='ignore', sparse_output=False))])

In [254]:
# Making a ohe_tuple for ColumnTransformer
ohe_tuple = ('categorical', ohe_pipe, ohe_cols)
ohe_tuple

('categorical',
 Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='constant')),
                 ('onehotencoder',
                  OneHotEncoder(handle_unknown='ignore', sparse_output=False))]),
 Index(['Item_Fat_Content', 'Outlet_Location_Type', 'Outlet_Type'], dtype='object'))

Numeric feature pipeline

In [255]:
# PREPROCESSING PIPELINE FOR NUMERIC DATA
# Save list of column names
num_cols = X_train.select_dtypes("number").columns
print("Numeric Columns:", num_cols)
# instantiate preprocessors
impute_median = SimpleImputer(strategy='median')
scaler = StandardScaler()
# Make a numeric preprocessing pipeline
num_pipe = make_pipeline(impute_median, scaler)
num_pipe

Numeric Columns: Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')


Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='median')),
                ('standardscaler', StandardScaler())])

In [256]:
# Making a numeric tuple for ColumnTransformer
num_tuple = ('numeric', num_pipe, num_cols)
num_tuple

('numeric',
 Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='median')),
                 ('standardscaler', StandardScaler())]),
 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
        'Outlet_Establishment_Year'],
       dtype='object'))

**Instantiate the ColumnTransformer**



In [258]:
# Instantiate with verbose_feature_names_out=False
col_transformer = ColumnTransformer([num_tuple, ord_tuple, ohe_tuple],
                                    verbose_feature_names_out=False)
col_transformer

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='median')),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('ordinal',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('ordinalencoder',
                                                  Ord...
                                                                              'Medium',
                                                                              'Small']])),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 ['Outlet_Size']),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Fat_Content', 'Outlet_Location_Type', 'Outlet_Type'], dtype='object'))],
                  verbose_feature_names_out=False)

**Fit the Transformer on Training Data Only**

In [259]:
# Fit on training data
col_transformer.fit(X_train)

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='median')),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('ordinal',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('ordinalencoder',
                                                  Ord...
                                                                              'Medium',
                                                                              'Small']])),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 ['Outlet_Size']),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Fat_Content', 'Outlet_Location_Type', 'Outlet_Type'], dtype='object'))],
                  verbose_feature_names_out=False)

**Transform Training and Test data**

In [260]:
# Transform the training data
X_train_processed = col_transformer.transform(X_train)
# Transform the testing data
X_test_processed = col_transformer.transform(X_test)

In [261]:
# View the processed training data & testing data
display(X_train_processed.head(),X_test_processed.head())

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
4776     0.819191        -0.712775  1.828109                   1.327849   
7510     0.558284        -1.291052  0.603369                   1.327849   
5828    -0.129561         1.813319  0.244541                   0.136187   
5327    -1.167259        -1.004931 -0.952591                   0.732018   
4810     1.530755        -0.965484 -0.336460                   0.493686   

      Outlet_Size  Item_Fat_Content_Low  Item_Fat_Content_Regular  \
4776    -0.287374                   1.0                       0.0   
7510    -0.287374                   0.0                       1.0   
5828    -0.287374                   0.0                       1.0   
5327     1.384048                   1.0                       0.0   
4810    -0.287374                   1.0                       0.0   

      Outlet_Location_Type_Tier 1  Outlet_Location_Type_Tier 2  \
4776                          0.0                          0.0   
7510                          0.0                          0.0   
5828                          1.0                          0.0   
5327                          0.0                          1.0   
4810                          0.0                          1.0   

      Outlet_Location_Type_Tier 3  Outlet_Type_Supermarket Type1  \
4776                          1.0                            0.0   
7510                          1.0                            0.0   
5828                          0.0                            1.0   
5327                          0.0                            1.0   
4810                          0.0                            1.0   

      Outlet_Type_Supermarket Type2  Outlet_Type_Supermarket Type3  \
4776                            1.0                            0.0   
7510                            1.0                            0.0   
5828                            0.0                            0.0   
5327                            0.0                            0.0   
4810                            0.0                            0.0   

      Outlet_Type_Supermarket Type4  
4776                            0.0  
7510                            0.0  
5828                            0.0  
5327                            0.0  
4810                            0.0

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
7503     0.332955        -0.776646 -0.998816                  -1.293807   
2957    -1.177932         0.100317 -1.585194                  -0.102145   
7031     0.380393        -0.482994 -1.595784                   0.136187   
1084    -0.009307        -0.415440  0.506592                  -1.532139   
856     -0.640701        -1.047426  0.886725                   0.732018   

      Outlet_Size  Item_Fat_Content_Low  Item_Fat_Content_Regular  \
7503    -1.958796                   1.0                       0.0   
2957     1.384048                   1.0                       0.0   
7031    -0.287374                   0.0                       1.0   
1084    -0.287374                   0.0                       1.0   
856      1.384048                   0.0                       1.0   

      Outlet_Location_Type_Tier 1  Outlet_Location_Type_Tier 2  \
7503                          0.0                          0.0   
2957                          1.0                          0.0   
7031                          1.0                          0.0   
1084                          0.0                          0.0   
856                           0.0                          1.0   

      Outlet_Location_Type_Tier 3  Outlet_Type_Supermarket Type1  \
7503                          1.0                            1.0   
2957                          0.0                            1.0   
7031                          0.0                            1.0   
1084                          1.0                            0.0   
856                           0.0                            1.0   

      Outlet_Type_Supermarket Type2  Outlet_Type_Supermarket Type3  \
7503                            0.0                            0.0   
2957                            0.0                            0.0   
7031                            0.0                            0.0   
1084                            0.0                            1.0   
856                             0.0                            0.0   

      Outlet_Type_Supermarket Type4  
7503                            0.0  
2957                            0.0  
7031                            0.0  
1084                            0.0  
856                             0.0